src/script.rs

```diff
@@ -534,44 +531,63 @@ impl From<Disabled> for u8 {
     }
 }
 
-/// When we parse a single [`Opcode`], the result is a bit complicated. It returns the remaining
-/// unparsed script code along with either a known [`Opcode`] or an unknown byte (which _may_ cause
-/// a script failure down the line).
+/// When writing scripts, we don’t want to allow bad opcodes, so `Opcode` doesn’t include them.
+/// However, when validating scripts, bad opcodes only cause a failure when they’re on an active
+/// branch, so this type allows us to hold onto the bad opcodes when parsing.
+pub enum PossiblyBad {
+    Good(Opcode),
+    Bad(Bad),
+}
+
 pub struct ParsedOpcode<'a> {
     /// The [`Result`] allows us to preserve unknown opcodes, which only trigger a failure if
     /// they’re on an active branch during interpretation.
-    pub opcode: Result<Opcode, Bad>,
+    pub opcode: PossiblyBad,
     pub remaining_code: &'a [u8],
 }
 
+/// “Strict” scripts disallow [`Bad`] opcodes, and so we collect any bad opcodes that we find, so we
+/// can report all of them, rather than just the first one.
+pub enum StrictError {
+    /// A script validation failure.
+    Script(ScriptError),
+    /// All of the [`Bad`] opcodes we found.
+    BadOpcodes(Vec<Bad>),
+}
+
 /** Serialized script, used inside transaction inputs and outputs */
 #[derive(Clone, Debug)]
 pub struct Script<'a>(pub &'a [u8]);
 
 impl Script<'_> {
     /// Fails on all [`Bad`] opcodes, not just ones on active branches.
-    pub fn parse_strict(&self) -> Result<Vec<Opcode>, Result<ScriptError, Bad>> {
-        // TODO: Collect _all_ [opcode::Bad] failures.
-        match self.parse().map(|script| script.into_iter().collect()) {
-            Err(op_err) => Err(Ok(op_err)),
-            Ok(Err(bad_ops)) => Err(Err(bad_ops)),
-            Ok(Ok(ops)) => Ok(ops),
-        }
+    pub fn parse_strict(&self) -> Result<Vec<Opcode>, StrictError> {
+        self.parse()
+            .map_err(StrictError::Script)
+            .and_then(|script| {
+                let (opcodes, bad_opcodes): (Vec<_>, Vec<_>) =
+                    script.into_iter().partition_map(|pb| match pb {
+                        PossiblyBad::Good(opcode) => Either::Left(opcode),
```

Comment:
I'm embarrassed to say that I'd never noticed https://docs.rs/itertools/latest/itertools/enum.Either.html before - we've used a different `either` crate elsewhere, but getting `Either` from `itertools` is definitely better.